src/linker/Linker.Steps/SealerStep.cs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Collections.Generic;
+using Mono.Cecil;
+
+namespace Mono.Linker.Steps
+{
+	public class SealerStep : BaseStep
+	{
+		public SealerStep ()
+		{
+		}
+
+		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		{
+			if (Annotations.GetAction (assembly) != AssemblyAction.Link)
+				return;
+
+			if (!Context.Optimizations.IsEnabled (CodeOptimizations.Sealer, assembly))
+				return;
+
+			foreach (var type in assembly.MainModule.Types)
+				ProcessType (type);
+		}
+
+		bool IsSubclassed (TypeDefinition type)
+		{
+			foreach (var a in Context.GetAssemblies ()) {
+				foreach (var s in a.MainModule.Types) {
+					if (s.BaseType != null && s.BaseType.IsTypeOf (type.Namespace, type.Name) == true)
+						return true;
+
+					if (s.HasNestedTypes) {
+						foreach (var ns in s.NestedTypes) {
+							if (ns.BaseType.IsTypeOf (type.Namespace, type.Name))
+								return true;
+						}
+					}
+				}
+			}
+			return false;
+		}
+
+		void ProcessType (TypeDefinition type)
+		{
+			//
+			// interface members are virtual (and we cannot change this)
+			//
+			if (type.IsInterface)
+				return;
+
+			//
+			// the code does not include any subclass for this type
+			//
+			if (!type.IsAbstract && !type.IsSealed && !IsSubclassed (type))
+				type.IsSealed = true;
```

Comment:
Would you be willing to move all type/method changes into protected virtual methods so that I can override them in our linker?  I like to wire up all changes the linker makes to some sort of diagnostic logging so that we have a way to view the changes the linker makes if we want to.

I'm thinking something like

```
protected virtual void MakeTypedSealed(TypeDefinition type)
{
    type.IsSealed = true;
}

protected virtual void MakeMethodFinal(MethodDefinition method)
{
   method.IsFinal = true;
}
```

Reply:
Done